### ISSUE TITLE
`fly set-pipeline` shows number in weird exponential notation, which was not changed

### ISSUE THREAD

**Issue by petergtz** [concourse/concourse#1407]
**Bug Report**

After our latest update to Concourse 3.3.2 (from 2.6) we found a minor regression when updating a pipeline with `fly set-pipeline`. Even though no change was made to the number in the following diff, it still shows up in this weird way. (the exponential notation was never used in the yml):

```diff
      - task: my-task
        file: mytask.yml
        params:
          ...
          KEEP_NUM_VMS: 0
-          PRIVATE_VLAN: 1.418073e+06
+          PRIVATE_VLAN: 1418073
```
(numbers changed for this report)

**Additional information:**

* Deployment type: BOSH
* Infrastructure/IaaS: vSphere

**Comment by jtarchie** [comment]
@petergtz, we are using a new YAML parser, which might has some stricter representation when it comes to YAML.

Besides the diff, did this affect the way your task ran? I believe it is the same value.

**Comment by petergtz** [comment]
> Besides the diff, did this affect the way your task ran? I believe it is the same value.

Yes, it's the same value. I'm using this as an ID and if it came into the task differently than intended it wouldn't work as ID. So it's not affecting the task.

It seems it's a pure diff issue. 

The problem is that it gets shown every time, even when I haven't changed that value which makes the diff a lot longer and more difficult to quickly review.

**Comment by jtarchie** [comment]
We saw this issue many months ago. It has to do with the YAML interpreting the type as the actual number in YAML and then comparing it to the string stored in the database.

Thanks for clarifying.

**Comment by petergtz** [comment]
Note that I also only see this for this 7 digit ID, not for other 6 digit IDs.

**Comment by petergtz** [comment]
> Besides the diff, did this affect the way your task ran? I believe it is the same value.

Yes, it's the same value. I'm using this as an ID and if it came into the task differently than intended it wouldn't work as ID. So it's not affecting the task.

It seems it's a pure diff issue. 

The problem is that it gets shown every time, even when I haven't changed that value which makes the diff a lot longer and more difficult to quickly review.

**Comment by XenoPhex** [comment]
Bumping this issue again as it's preventing the Cloudfoundry VAT team from using the tracker resource:
```
resources:
  resource cf-cli-tracker has changed:
  name: cf-cli-tracker
  type: tracker
  source:
-   project_id: 2.196383e+06
+   project_id: 2196383
    token: [REDACTED]
    tracker_url: https://www.pivotaltracker.com

apply configuration? [yN]:
```
Thus causing this error in our pipeline:
![error reading request: json: cannot unmarshal number into Go struct field Source.project_id of type string](https://user-images.githubusercontent.com/291961/46965407-9ed48b00-d05f-11e8-92b9-39da65828a72.png)

If someone can point out all the places that need the change, I can try taking a whack at fixing this in my free time.

**Comment by jtarchie** [comment]
@XenoPhex: The error in your screenshot actually states that it needs a string. Can you put the `project_id` value into strings (`""`)?

**Comment by XenoPhex** [comment]
That's both with and without quotes in our pipeline yaml.

**Comment by marco-m** [comment]
Hello, I think I cannot reproduce this with Concourse 5.0. If I am correct, this can issue can be closed.

Consider the following pipeline

```
jobs:
- name: hello-world
  plan:
  - task: say-hello
    config:
      platform: linux
      image_resource:
        type: docker-image
        source: {repository: alpine}
      run:
        path: echo
        args: ["Hello, world!"]
      params:
        vat-public-tracker-project-id: 2196383
```

Setting it:

```
fly -t prod-developers set-pipeline -p deleteme-marco -c issue-1407-pipeline.yml
jobs:
  job hello-world has been added:
+ name: hello-world
+ plan:
+ - task: say-hello
+   config:
+     platform: linux
+     image_resource:
+       type: docker-image
+       source:
+         repository: alpine
+     params:
+       vat-public-tracker-project-id: "2196383"
+     run:
+       path: echo
+       args:
+       - Hello, world!

apply configuration? [yN]: y
pipeline created!
```

Running it works correctly.

Resetting it says nothing as changed (as expected):

```
fly -t prod-developers set-pipeline -p deleteme-marco -c issue-1407-pipeline.yml
no changes to apply
```

**Comment by DanielJonesEB** [comment]
I hit this issue with a customer on Concourse 5.1.0. 

```
      TF_VAR_folder_id: "((folder_id))"
```

...where `folder_id` is provided as a command-line var with a value of `368574885298`.

**Comment by petergtz** [comment]
Unfortunately, I can't check right now, but I'll try to check if this is still an issue next week.

**Comment by vito** [comment]
@DanielJonesEB How is that value (123456789012) being provided on the CLI? If it's with `-y` or `-l`, those both parse as YAML, and that value will parse as a floating point number. You may have to take care to quote the value, i.e. `-y foo="123456789012"`.

In general, if you're passing a numeric value around I would be careful to quote it everywhere, including vars files passed with `-l`, otherwise it may be impossible to preserve its exact representation. I believe YAML and JSON both treat all numbers as floating point when parsed into a nonspecific type, which is what `source:` and `params:` are.

**Comment by vito** [comment]
There was also a case where we took special care to handle floating point values passed to task params and stringify them:

https://github.com/concourse/concourse/blob/aab8f4a85d0146a3b2ae2e166c83f459afecb4e3/atc/decode_hook.go#L88-L94

...but I think we should get rid of this and just make sure people quote their values appropriately.

**Comment by petergtz** [comment]
I do still hit this issue with version 5.3.0:
```diff
$ fly --version
5.3.0

$ fly -t flintstone set-pipeline -c number-test.yml -p number-test
jobs:
  job test has changed:
  name: test
  plan:
  - task: some-tesk
    file: some-file
    params:
-     some: 1.234567e+06
+     some: 1234567
```

@vito 
> ...but I think we should get rid of this and just make sure people quote their values appropriately.

I guess that would make sense, as it is not obvious that this is the way to fix it. How would you make sure people quote their values?

Would it make sense in addition to treat `param` values always as strings? Environment variables have no notion of types. That could help to avoid this problem in the first place.

**Comment by vito** [comment]
@petergtz Here's one strategy: for `task`, introduce `env:` instead, and require all values to be strings. We can then deprecate `task:` + `params:`, preserving today's funky coercion behavior until we eventually remove it. Task configs would also switch to `env:` instead of `params:`, but we'd allow both with a deprecation warning for backwards-compatibility.

The main reason I want to do this is to stop overloading `params:`. For `get` and `put`, it's arbitrary config (not strings) to pass to the resource. For tasks, it's env vars. Kind of confusing to give them both the same name.

With concourse/rfcs#7 in mind, we may eventually want `params:` to actually behave the same as with `get`/`put`, and we can't do that if it's already being used for env vars. It's kind of an awkward transition, (`params` for env -> deprecated `params` -> `params` for task params), but we'd at least be consistent across `get`/`put`/`task` in the end.

Failing all that, we could just add a validation warning that all values are strings, and turn it into an error after folks have had enough time to transition.

**Comment by petergtz** [comment]
@vito Sounds good to me. So it seems to me that the work for all your suggestions is tracked somewhere else, which seems to suggest to leave the **workaround/solution** here for anyone to quickly find it:

**Values should be quoted appropriately to makes sure they're not interpreted as numbers. Example:**
```yaml
jobs:
  job test has changed:
  name: test
  plan:
  - task: some-tesk
    file: some-file
    params:
      some: "1234567"
```

and close this issue. Feel free to do so, if you agree.

**Comment by vito** [comment]
@petergtz Nothing tracking it yet. So thanks for making the work-around easy to find! I'll close this once I figure out where to put the `env:` stuff in our [roadmap](https://github.com/concourse/concourse/projects/21). :thinking:

**Comment by vito** [comment]
Whaat, I moved this into a project so it wouldn't be closed. Guess I have to remove the label too.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
